var/spack/repos/builtin/packages/exago/package.py

```diff
@@ -124,14 +124,10 @@ class Exago(CMakePackage, CudaPackage, ROCmPackage):
             when="+{0} build_type=RelWithDebInfo".format(pkg[1]),
         )
 
-    depends_on(
-        "{0} build_type=Release".format("hiop+ginkgo ^ginkgo"),
-        when="+{0} build_type=Release".format("hiop ^hiop+ginkgo"),
-    )
-    depends_on(
-        "{0} build_type=Debug".format("hiop+ginkgo ^ginkgo"),
-        when="+{0} build_type=RelWithDebInfo".format("hiop ^hiop+ginkgo"),
-    )
```

Comment:
Not really sure what this was meant to do, so I simplified it. Also, kind of weird to use "format"... with literal strings as arguments.

Reply:
@cameronrutherford @pelesh @ryandanehy Let me know if this change is fine, or if you need a different semantic.

Reply:
It looks like this was trying to propagate the `build_type` and that this is more concerned with adding a dependency on `ginkgo` - expanding the second one:

```
depends_on("hiop+ginkgo ^ginkgo build_type=Debug",
  when="+hiop ^hiop+ginkgo build_type=RelWithDebInfo")
```

so another interpretation could be:

```
with when("+hiop ^hiop+ginkgo"):
  depends_on("ginkgo build_type=Debug", when="^hiop build_type=RelWIthDebInfo)
```

In words: "I need ginkgo when I need hiop and hiop needs ginkgo; also build ginkgo in debug mode if hiop was built in debug mode"

(there's already a dependency on `hiop` added up around line 106)

Reply:
@jaelynlitz and I are in agreement with @scheibelp interpretations we already use the simple depends on earlier in the code when we constrain hiop version number lines 136-140. This snippet is trying to propagate build type to packages. So we like @scheibelp suggestion of

``
with when("+hiop ^hiop+ginkgo"):
  depends_on("ginkgo build_type=Release", when="^hiop build_type=Release)

with when("+hiop ^hiop+ginkgo"):
  depends_on("ginkgo build_type=Debug", when="^hiop build_type=RelWIthDebInfo)``

Note: we still have to test to make sure.

Reply:
After some fiddling locally -  this seems to give the correct behavior on our side, not sure if this will pass on the audit

```
     with when("+hiop build_type=Release ^hiop+ginkgo"):                         
       depends_on("hiop build_type=Release")                                     
       depends_on("ginkgo build_type=Release", when="^hiop build_type=Release")  
                                                                                 
     with when("+hiop build_type=Debug ^hiop+ginkgo"):                           
       depends_on("hiop build_type=RelWithDebInfo")                                                                                                                                
       depends_on("ginkgo build_type=Debug", when="^hiop build_type=RelWithDebInfo")
```

Reply:
See https://github.com/spack/spack/pull/41428/commits/68bae59b80b6d33832bd3e6bcf973abc8b239a55 I avoided things like:
```
depends_on("foo", when="^foo")
```
cause they are tautological and might add nondeterminism by giving clingo the "choice" to put an extra node. I tried to propagate the `build_type` like you do here, though I think this is better done using configuration.

Reply:
`exago` is failing to concretize in pipelines. I'm looking at what is the issue, but cannot reproduce locally so far.

Reply:
@jaelynlitz @ryandanehy There are a couple of points that are still not clear to me: 
1. Does `exago+hiop` require `hiop+ginkgo` or is `hiop~ginkgo` fine too?
2. If `hiop~ginkgo` is fine, do I understand correctly from https://github.com/spack/spack/pull/41428#discussion_r1416305778 that you want to forward the build types only if `hiop+ginkgo` is used?

Reply:
1. I believe `hiop~ginkgo` is viable, but yes `exago+hiop+ginkgo` only works with the `glu_experimental` branch of ginkgo. 
2. Our goal is to in general propagate the same or similar `build_type` down to our dependencies and their dependencies. Like in this loop of packages [L112-125](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/exago/package.py#L120). If that syntax looks okay, we could probably just move `hiop+ginkgo` to be included there?

Reply:
https://github.com/pnnl/ExaGO/issues/50 - tracks us upgrading from the old ginkgo version. Thanks all for the help, and I think this change works well